### PR TITLE
Revert #868

### PR DESCRIPTION
**Pull request by jplatte**
… and a few things that built on it.

### CODE REVIEW

**Review by poljar — APPROVED** [review]
Looks good, the autojoin example works again as expected. Thanks.

File: crates/matrix-sdk/src/room/common.rs [hunk 1]
```diff
@@ -107,73 +101,22 @@ impl Common {
...
-    }
+    pub(crate) async fn join(&self) -> Result<()> {
+        let request = join_room_by_id::v3::Request::new(self.inner.room_id());
+        let _response = self.client.send(request, None).await?;
 
-    /// Get the inner client saved in this room instance.
-    ///
-    /// Returns the client this room is part of.
-    pub fn client(&self) -> Client {
-        self.client.clone()
```
Comment (FlixCoder) [inline crates/matrix-sdk/src/room/common.rs:176]: Why does it remove the `client()` function? (And it keeps its doc comment on join instead)

Comment (jplatte) [reply crates/matrix-sdk/src/room/common.rs:176]: Git auto-resolving conflicts gone wrong, I guess. Thanks for noticing!

File: testing/matrix-sdk-integration-testing/src/tests/repeated_join.rs [hunk 2]
```diff
@@ -1,144 +0,0 @@
...
-        println!("Joining..");
-        let room =
-            karl.get_invited_room(room_id).expect("karl has the room").accept_invitation().await?;
-        println!("Done");
-        let membership = room.get_member_no_sync(&karl_id).await?.expect("karl joined");
-        assert_eq!(*membership.membership(), MembershipState::Join);
-
-        assert!(karl.get_invited_room(room_id).is_none());
-        assert!(karl.get_joined_room(room_id).is_some());
-        assert!(karl.get_left_room(room_id).is_none());
```
Comment (FlixCoder) [inline testing/matrix-sdk-integration-testing/src/tests/repeated_join.rs:77]: Removing my whole test without replacement agian elaves open the possibility for rooms to be in multiple states in the SDK after re-joining.. :(

Comment (poljar) [reply testing/matrix-sdk-integration-testing/src/tests/repeated_join.rs:77]: The PR removes a bit more than needed, but untangling all of this isn't trivial. Don't worry, we'll come back to it by the end of the week.

Comment (FlixCoder) [reply testing/matrix-sdk-integration-testing/src/tests/repeated_join.rs:77]: Alright, thanks

**Comment by timthelion** [comment]
Is there some sort of context that I'm missing as to why this was reverted? I understand that an example, `autojoin` was broken, but was there some fundamental design flaw that hampered this approach? Is there something I can do to restore the functionality that was reverted?

**Comment by gnunicorn** [comment]
@timthelion ﻿﻿#1041 provides some context - it was the integration test for that feature and it showed, that we have an inherent racy-ness because of the inner architecture of our code, including some bad side effects like inconsistent state and potential data loss. The core problem is that the join and leave commands from the server don't allow us to fully update the local state because they don't provide enough information. Thus the solution was to attach to the sync-poll and wait for the proper events to arrive - but that is inherently racy, as this needs to happen out of bounds and might as well has already happened when we start listening to the event. The initial design would set up the listener before doing the send to prevent that but that would lead to an endless deadlock if using in an event_handler (which is a very common pattern), and even there we only _assume_ certain events belong to our send, but have no way of verifying it.

The proper design would be to have a another incomplete-room-state where we prefill the room with the information we can gather safely but make clear that they have not been confirmed by the server. Thus we can return a probably-joined-room right from the `join` and are not running into races or inconsistencies. That however, requires a bigger internal refactoring (and potential data migration) we haven't scheduled yet.

**Comment by timthelion** [comment]
Aha, that makes sense. I had mentioned the possible race ("Thus the solution was to attach to the sync-poll and wait for the proper events to arrive - but that is inherently racy, as this needs to happen out of bounds and might as well has already happened ") in my own PR but had not confirmed it. Sorry to see that I was correct... I had an incomplete version where I set up the sync/poll event handler and THEN checked to see if the room already exists and if not, continued to wait. Is that not a valid solution?

**Comment by gnunicorn** [comment]
> I had an incomplete version where I set up the sync/poll event handler and THEN checked to see if the room already exists and if not, continued to wait. Is that not a valid solution?

The problem is any inherent await we do _before_ sending (which is sort-of the only way we can prevent the race directly), means we create a food-gun if you then did that `.await` on it from any `event_handler` as this will logically deadlock: it won't send the next request until this await is finished, which itself awaits for results from the next sync response, which never happens because of the await.

Considering that this is a _super common_ use case (see an invite, autojoin the room), this is a food gun we are not willing to introduce.

**Comment by timthelion** [comment]
And can we re-emit the next sync response? In old event base GUI programming like in GTK or QT there is the concept that an event is either consumed or re-emitted, could a similar concept prevent that deadlock?

**Comment by gnunicorn** [comment]
(we should probably discuss this somewhere else? How about the issue on #1041 ?)

**Pull request merged**